addons/bus/websocket.py

```diff
@@ -0,0 +1,863 @@
+            # the other end does not close the socket properly.
+            self._socket.settimeout(1)
+            while self._socket.recv(4096):
+                pass
+        self._selector.unregister(self._socket)
+        self._selector.close()
+        self._socket.close()
+        self.state = ConnectionState.CLOSED
+        type(self)._instances.remove(self)
+        self._trigger_lifecycle_event(LifecycleEvent.CLOSE)
+
+    def _handle_control_frame(self, frame):
+        if frame.opcode is Opcode.PING:
+            self._enqueue_pong_frame(frame.payload)
+        elif frame.opcode is Opcode.CLOSE:
+            self.state = ConnectionState.CLOSING
+            self._close_received = True
+            code, reason = CloseCode.CLEAN, None
+            if len(frame.payload) >= 2:
+                code = struct.unpack('!H', frame.payload[:2])[0]
+                reason = frame.payload[2:].decode('utf-8')
+            elif frame.payload:
+                raise ProtocolError("Malformed closing frame")
+            if not self._close_sent:
+                self._enqueue_close_frame(code, reason)
+            else:
+                self._terminate()
+
+    def _handle_transport_error(self, exc):
+        """
+        Find out which close code should be sent according to given
+        exception and call `self.disconnect` in order to close the
+        connection cleanly.
+        """
+        if self.state is ConnectionState.CLOSED:
+            return
+        code, reason = CloseCode.SERVER_ERROR, str(exc)
+        if isinstance(exc, (ConnectionClosed, OSError)):
+            code = CloseCode.ABNORMAL_CLOSURE
+        elif isinstance(exc, (ProtocolError, InvalidCloseCodeException)):
+            code = CloseCode.PROTOCOL_ERROR
+        elif isinstance(exc, UnicodeDecodeError):
+            code = CloseCode.INCONSISTENT_DATA
+        elif isinstance(exc, PayloadTooLargeException):
+            code = CloseCode.MESSAGE_TOO_BIG
+        elif isinstance(exc, RateLimitExceededException):
+            code = CloseCode.TRY_LATER
+        if code is CloseCode.SERVER_ERROR:
+            reason = None
+            _logger.error(exc, exc_info=True)
+        self.disconnect(code, reason)
+
+    def _limit_rate(self):
+        """
+        This method is a simple rate limiter designed not to allow
+        more than one request by `RL_DELAY` seconds. `RL_BURST` specify
+        how many requests can be made in excess of the given rate at the
+        begining. When requests are received too fast, raises the
+        `RateLimitExceededException`.
+        """
+        now = time.time()
+        if len(self._incoming_frame_timestamps) >= type(self).RL_BURST:
+            elapsed_time = now - self._incoming_frame_timestamps[0]
+            if elapsed_time < type(self).RL_DELAY * type(self).RL_BURST:
+                raise RateLimitExceededException()
+        self._incoming_frame_timestamps.append(now)
+
+    @classmethod
+    def _on_server_stop(cls):
```

Comment:
```suggestion
    @classmethod
    @CommonServer.on_stop
    def _kick_all(cls):
```

Reply:
And remove the dangling `CommonServer.on_stop(...)` at the end of the file

Reply:
But we should change the `on_stop` implementation so that it returns the function

```py
    def on_stop(self, func):
        """ Register a cleanup function to be executed when the server stops """
        self._on_stop_funcs.append(func)
        return func
```

Reply:
@Julien00859, I'm not sure this is a good idea.  This morph the `on_stop` method into a decorator, but we won't be able to pass `self`/`cls` arguments to the registered methods. All in all, I don't think this is worth implementing.

Reply:
> but we won't be able to pass self/cls arguments to the registered methods

The function is decorated with `@classmethod` so the `cls` argument should be passed fine

Reply:
But it is fine this way, I was most bothered by the function name :D

Reply:
`@classmethod` decorates the result of `@CommonServer._on_stop` which means `on_stop` is actually registering an unbound method (so `cls` won't be pass to the registered callback).